src/ExtranetBundle/LoginByTransferToken/TransferTokenService.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace ChameleonSystem\ExtranetBundle\LoginByTransferToken;
+
+use ChameleonSystem\CoreBundle\Interfaces\TimeProviderInterface;
+
+class TransferTokenService implements TransferTokenServiceInterface
+{
+    /** @var TimeProviderInterface */
+    private $timeProvider;
+
+    /** @var string */
+    private $secret;
+
+    /** @var string */
+    private $algorithm;
+
+    public function __construct(
+        TimeProviderInterface $timeProvider,
+        string $secret,
+        string $algorithm
+    ) {
+        $this->timeProvider = $timeProvider;
+        $this->secret = $secret;
+        $this->algorithm = $algorithm;
+    }
+
+    /** {@inheritdoc} */
+    public function createTransferTokenForUser(
+        string $userId,
+        int $expiresAfterSeconds
+    ): string {
+        $salt = md5(random_bytes(16));
+        $expires = $this->timeProvider->getUnixTimestamp() + $expiresAfterSeconds;
+        $json = json_encode(compact('userId', 'expires', 'salt'));
+        $encrypted = openssl_encrypt(
+            $json,
+            $this->algorithm,
+            $this->secret,
+            0,
+            $this->initializationVector()
+        );
+
+        return base64_encode($encrypted);
+    }
+
+    /** {@inheritdoc} */
+    public function validateTransferToken(string $token): ?string
```

Comment:
the name of this method does not imply that it will return a userid on success. Sure, it does validate the token in the process but a caller might expect a boolean return value.

Reply:
Fair enonugh - I renamed it to `getUserIdFromTransferToken` which should communicate that a bit better.